src/Workspaces/Remote/Core/Services/Options/FileWatcher.cs

```diff
@@ -0,0 +1,127 @@
+
+        public void Dispose()
+        {
+            var watchers = Interlocked.Exchange(ref _watchers, null);
+            if (watchers == null)
+            {
+                return;
+            }
+
+            ConventionFileChanged = null;
+            ContextFileMoved = null;
+        }
+
+        public void StartWatching(string fileName, string directoryPath)
+        {
+            FileSystemWatcher watcher = null;
+            var updated = ImmutableInterlocked.Update(
+                ref _watchers,
+                watchers =>
+                {
+                    if (watchers is null)
+                    {
+                        watcher?.Dispose();
+                        throw new ObjectDisposedException(nameof(FileWatcher));
+                    }
+
+                    if (!watchers.TryGetValue((fileName, directoryPath), out var existingWatcher))
+                    {
+                        if (watcher == null)
+                        {
+                            watcher = new FileSystemWatcher(directoryPath, fileName);
+                        }
+
+                        return watchers.Add((fileName, directoryPath), watcher);
+                    }
+
+                    return watchers;
+                });
+
+            if (updated)
+            {
+                FileSystemEventHandler handler = (sender, e) => HandleFileSystemEvent(fileName, directoryPath, e);
+                watcher.Changed += handler;
+                watcher.Created += handler;
+                watcher.Deleted += handler;
+                watcher.Renamed += HandleRenamedEvent;
+            }
+            else
+            {
+                watcher?.Dispose();
+            }
+        }
+
+        private void HandleFileSystemEvent(string fileName, string directoryPath, FileSystemEventArgs e)
+        {
+            if (e.ChangeType.HasFlag(WatcherChangeTypes.Created))
+            {
+                ConventionFileChanged?.Invoke(this, new ConventionsFileChangeEventArgs(fileName, directoryPath, ChangeType.FileCreated));
+            }
+
+            if (e.ChangeType.HasFlag(WatcherChangeTypes.Changed))
+            {
+                ConventionFileChanged?.Invoke(this, new ConventionsFileChangeEventArgs(fileName, directoryPath, ChangeType.FileCreated));
+            }
+
+            if (e.ChangeType.HasFlag(WatcherChangeTypes.Deleted))
+            {
+                ConventionFileChanged?.Invoke(this, new ConventionsFileChangeEventArgs(fileName, directoryPath, ChangeType.FileDeleted));
+            }
+        }
+
+        private void HandleRenamedEvent(object sender, RenamedEventArgs e)
+        {
+            ContextFileMoved?.Invoke(this, new ContextFileMovedEventArgs(e.OldFullPath, e.FullPath));
+        }
+
+        public void StopWatching(string fileName, string directoryPath)
+        {
+            FileSystemWatcher watcher = null;
+            var updated = ImmutableInterlocked.Update(
```

Comment:
Ditto on "why not a lock?" This is tricky to follow mostly because it's violating the documented rule that the lambda given shouldn't have side-effects. It works, but it took me minutes to confirm it vs. seconds otherwise. :smile: